### PR TITLE
Templates

### PR DESCRIPTION
**Pull request by JW-Vinayak**
Have added functionality to fetch templates list and to create app, replace the template. Please review and provide your feedback.

### CODE REVIEW

**Comment by argelius** [comment]
Thanks @JW-Vinayak !

I will take a look at it tomorrow.

File: src/monaca.js [hunk 1]
```diff
@@ -1182,5 +1193,168 @@
     return deferred.promise;
   };
 
+
+  Monaca.prototype.getTemplates = function(){
+      var deferred = Q.defer();
```
Comment (argelius) [inline src/monaca.js]: Fix indentation.

File: src/monaca.js [hunk 2]
```diff
@@ -1182,5 +1193,168 @@
...
+          list = [];
+          list.push({
+            name : 'Minimal Cordova Template',
+            path : null
+          });
+          var files = fs.readdirSync(dir);
+          files.forEach(function(file){
+              if (file.match(/\.zip$/g)) {
+                  list.push({
+                      name: file.replace(/\.zip/g, '').replace(/onsen_/g, '').replace(/_/g, ' '),
```
Comment (argelius) [inline src/monaca.js]: Can you capitalize as well and not remove "onsen" ?

onsen_master_detail.zip -> Onsen Master Detail

File: src/monaca.js [hunk 3]
```diff
@@ -1182,5 +1193,168 @@
...
+                      path: path.join(dir, file)
+                  });
+              }
+          });
+          deferred.resolve(list);
+      }
+      catch(e){
```
Comment (argelius) [inline src/monaca.js]: Please use whitespaces before {

```
catch(e) {
```

File: src/monaca.js [hunk 4]
```diff
@@ -1182,5 +1193,168 @@
...
+        deferred.reject(err);
```
Comment (argelius) [inline src/monaca.js]: This variable doesn't exist.

File: src/monaca.js [hunk 5]
```diff
@@ -430,9 +438,12 @@
                     session: headers['x-monaca-param-session']
                   };
 
+                  console.log(JSON.stringify(this.tokens));
```
Comment (argelius) [inline src/monaca.js]: Can you remove debug prints like this one?

File: src/monaca.js [hunk 6]
```diff
@@ -1182,5 +1193,168 @@
...
+          console.log("started compiling list of templates.");
+          var dir = path.join(__dirname, '..', 'templates'),
+          list = [];
+          list.push({
+            name : 'Minimal Cordova Template',
+            path : null
+          });
+          var files = fs.readdirSync(dir);
+          files.forEach(function(file){
+              if (file.match(/\.zip$/g)) {
```
Comment (argelius) [inline src/monaca.js]: Why global flag? I think `test` is better when you're just looking for a match:

`/\.zip$/.test('onsen_master_detail.zip'); // true`

File: src/monaca.js [hunk 7]
```diff
@@ -390,14 +398,14 @@
         if (clientId) {
           form.clientId = clientId;
         }
-
+        console.log("sending login request")
```
Comment (argelius) [inline src/monaca.js]: Missing semicolon. Please run jshint on files.

File: src/monaca.js [hunk 8]
```diff
@@ -1182,5 +1193,168 @@
...
+          });
+          deferred.resolve(list);
+      }
+      catch(e){
+        deferred.reject(err);
+      }
+      return deferred.promise; 
+  }
+
+  Monaca.prototype.createApp = function(template, appname, workingDir){
```
Comment (argelius) [inline src/monaca.js]: Add jsdoc comments for public methods. If you think they should only be used internally please prefix method name with underscore.

File: src/monaca.js [hunk 9]
```diff
@@ -1182,5 +1193,168 @@
...
+                   console.log("process exit code is " + code);
+                   done(code);
+              }
+          });
+        } else
+          done();
+      }
+    ],
+    function(err, results){
+        if(err)
```
Comment (argelius) [inline src/monaca.js]: Please always use brackets for if statements.

**Comment by argelius** [comment]
I've reviewed the pull request.

Please add jsdoc comments, fix indentation, etc.
Also, I'd like you to add tests for the methods.

**Comment by JW-Vinayak** [comment]
Thank you Andreas. I will work on these points.

**Pull request merged**